extensions/machine-learning-services/src/test/queryRunner.test.ts

```diff
@@ -0,0 +1,303 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as azdata from 'azdata';
+import { ApiWrapper } from '../common/apiWrapper';
+import * as TypeMoq from 'typemoq';
+import * as should from 'should';
+import { QueryRunner } from '../common/queryRunner';
+
+interface TestContext {
+
+	apiWrapper: TypeMoq.IMock<ApiWrapper>;
+	queryProvider: azdata.QueryProvider;
+}
+
+function createContext(): TestContext {
+	return {
+		apiWrapper: TypeMoq.Mock.ofType(ApiWrapper),
+		queryProvider: {
+			providerId: '',
+			cancelQuery: undefined,
```

Comment:
@andresse@microsoft.com  is there anything change about our build? I used to be able to do this before and only implement the one I need in a interface. but now it gives me error that for example  cancelQuery cannot be undefined. what's the easiest way to get around this error if I don't care about this item in my test?

Reply:
That's because of the strict null checking that's [enabled by default](https://github.com/microsoft/azuredatastudio/blob/master/extensions/shared.tsconfig.json#L5). Just define stub functions that do nothing instead of undefined if you don't need any actual implementation.

Reply:
shouldn't that be enabled only for code not test? 

---
In reply to: [357768087](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r357768087) [](ancestors = 357768087)

Reply:
even If I create the stub, I have to implement the function to return something and that also cannot be undefined so I have to implement the result also. the chain goes on for ever. it's really annoying ... 

---
In reply to: [357770149](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r357770149) [](ancestors = 357770149,357768087)

Reply:
I mean this should be a one-time thing and once we have one made then other extensions can use that as well so it seems worth the effort. Strict null checking is very much a desirable thing so it's worth a bit of extra work now to keep it enabled.

Reply:
And we can't selectively enable this for parts of the code - it's a compiler option so unless your test code was somehow being compiled separately from the main product code you have to either have it or not.

Reply:
it's ok thanks, was not used to it but now I kind of know how to handle it

---
In reply to: [357776855](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r357776855) [](ancestors = 357776855)

Reply:
You can also just add a `!` to is and it will ignore the error `cancelQuery: undefined!`. but obviously only do this for test code.

Reply:
sure I fixed all the errors. all good now. thanks

---
In reply to: [357776158](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r357776158) [](ancestors = 357776158)

Reply:
I'll remember that for next time. thanks

---
In reply to: [357793384](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r357793384) [](ancestors = 357793384)